Clarify SDK behavior for Instrument Advisory Parameter (#4389)
open-telemetry/opentelemetry-specification: specification/metrics/sdk.md

      View sets an asynchronous instrument to use the [Explicit bucket
      histogram aggregation](#explicit-bucket-histogram-aggregation)) the
      implementation SHOULD emit a warning and proceed as if the View did not
      exist. If any configuration is provided via View and [Instrument advisory
      parameters](#instrument-advisory-parameters), then the View configuration

>       MUST take precedence.

Reviewer: Suppose the view sets the aggregation to be explicit bucket histogram and the advice says that the explicit bucket boundaries should be `[1, 5, 10]`. 

Does this language suggest that the output metric should be: 1. explicit bucket histogram with `[1, 5, 10]` or 2. explicit bucket histogram with default boundaries?

I believe it should be option 2, since if option 1 was defined it would mean that there is some sort of merging of the buckets from advisory parameters with the aggregation from the view.

Reviewer 2: +1, I think it should be option 2.

Author: Yes 2 is the expected behavior. There is no "merge" advice+view behavior today. And this PR is clarifying/documenting existing behavior only.

Do either of you suggest we should explicitly mention this?

Author: Resolving. If anyone feels the need to explicitly clarify this, please reopen.

### Instrument advisory parameters

> **Status**: [Stable](../document-status.md)

Reviewer 3: (since the Development status "Attributes" section is a sub-section of this)
```suggestion
**Status**: [Stable](../document-status.md), except where otherwise specified
```

> **Status**: [Stable](../document-status.md), except where otherwise specified

Reviewer 4: nit, could you please update PR title and/or add a changelog entry that explicit bucket boundaries are now stable in the SDK?

Author: it was stable already, but it was missed to update this section when it got stabilized. That's why I put this as merely editorial change...
(PR desc says this fixes https://github.com/open-telemetry/opentelemetry-specification/issues/4317 to cover this..)
Let me know if this is good enough.

Reviewer 5: (_Probably_ #4317 should be addressed before - to make clear that either the SDK portion is stable or _not_ yet)

Author: @carlosalberto Can you clarify more? 
SDK part is stable, just missed to update the status in the spec (that is my understanding, happy to correct if that was not the case.) This PR is fixing that.

first-seen advisory parameters and log an appropriate error as described in
[duplicate instrument registrations](#duplicate-instrument-registration).

If a [View](#view) is configured with the same settings as advisory parameters,
the View MUST take precedence over the advisory parameters.

#### Instrument advisory parameter: `ExplicitBucketBoundaries`

**Status**: [Stable](../document-status.md)

This parameter is applicable when the [Explicit Bucket
Histogram](#explicit-bucket-histogram-aggregation) aggregation is utilized.

`ExplicitBucketBoundaries` (`double[]`) specifies the recommended set of bucket
boundaries to use during the [Explicit Bucket
Histogram](#explicit-bucket-histogram-aggregation) aggregation. If the user has
configured a View with specific bucket boundaries, those boundaries should take
precedence. In the absence of such a configuration, the advisory parameter
should be used. If neither is provided, the default bucket boundaries must be

> applied.

Author: the last sentence may be too obvious that it may be omitted...But I am okay with being more explicit.

first-seen advisory parameters and log an appropriate error as described in
[duplicate instrument registrations](#duplicate-instrument-registration).

If both a [View](#view) and advisory parameters specify the same aspect of the

> [Stream configuration](#stream-configuration), the settings defined by the View

Reviewer 3: (to help reinforce that it's not all the settings)
```suggestion
[Stream configuration](#stream-configuration), the setting defined by the View
```

MUST take precedence over the advisory parameters.

#### Instrument advisory parameter: `ExplicitBucketBoundaries`

> **Status**: [Stable](../document-status.md)

Reviewer: This is redundant, since the parent section is already "stable, exception where otherwise specified"
```suggestion
```

This advisory parameter applies when the [Explicit Bucket
Histogram](#explicit-bucket-histogram-aggregation) aggregation is used.

If a matching View specifies Explicit Bucket Histogram aggregation with custom
bucket boundaries, those boundaries take precedence. If no View matches, or if a
matching View selects the default aggregation, the ExplicitBucketBoundaries

> advisory parameter must be used. If neither is provided, the default bucket

Reviewer 3: ```suggestion
advisory parameter MUST be used. If neither is provided, the default bucket
```

> Histogram](#explicit-bucket-histogram-aggregation) aggregation. If the user has
> configured custom bucket boundaries via View(s), those boundaries take
> precedence. If no View is configured, the advisory parameter should be used. If
> neither is provided, the default bucket boundaries must be applied.

Reviewer: I find this language ambiguous. Per the conversation [here](https://github.com/open-telemetry/opentelemetry-specification/pull/4389/files#r1960085312), I believe that `ExplicitBucketBoundaries` should only apply if the default aggregation is used. If a view sets Explicit bucket histogram aggregation the `ExplicitBucketBoundaries` parameter should be ignored even if the view doesn't specify bucket boundaries. 

```suggestion
Histogram](#explicit-bucket-histogram-aggregation) aggregation. If a matching View specifies Explicit bucket histogram aggregation, then the `ExplicitBucketBoundaries` advisory parameter is ignored. If no View matches, or if a matching view selects the default aggregation, the advisory parameter must be used. 
```

Reviewer 6: Sounds good -- so, if the view is set, the advisory param is ignored as a hard rule -- if you configure explicit boundaries w/o boundaries, you'll get the specified default and not the advisory param.

Author: PTAL https://github.com/open-telemetry/opentelemetry-specification/pull/4389/commits/dfca5ced9dc921dcd787f51519602a3f59f3f816

Reviewer: The language in https://github.com/open-telemetry/opentelemetry-specification/commit/dfca5ced9dc921dcd787f51519602a3f59f3f816 implies something different:

> If a matching View specifies Explicit Bucket Histogram aggregation with custom
bucket boundaries, those boundaries take precedence.

It implies that the advisory parameter is ignored only if a maching view specifies Explicit bucket histogram aggregation AND bucket boundaries. As @jmacd summarized:

>  so, if the view is set, the advisory param is ignored as a hard rule -- if you configure explicit boundaries w/o boundaries, you'll get the specified default and not the advisory param.

Author: @jack-berg Thanks. I made another tweak to make it clear that - if a view is set (with or without view specifying custom buckets), advice is ignored.

#### Instrument advisory parameter: `Attributes`

**Status**: [Development](../document-status.md)

This advisory parameter applies to all aggregations.

`Attributes` (a list of [attribute keys](../common/README.md#attribute))
specifies the recommended set of attribute keys for measurements aggregated to
produce a metric stream.

If the user has provided attribute keys via View(s), those keys take precedence.
If no View is configured, or if a matching view does not specify attribute keys,

> the advisory parameter should be used. If neither is provided, all attributes

Reviewer 3: any reason for SHOULD vs MUST which is used above for similar statement related to ExplicitBucketBoundaries?
```suggestion
the advisory parameter should be used. If neither is provided, all attributes
```

> If no View is configured, the advisory parameter should be used. If neither is

Reviewer: ```suggestion
If no View is configured, or if a matching view does not specify attribute keys, the advisory parameter should be used. If neither is
```

> provided, all attributes must be retained.

Reviewer 3: What if the user has provided a view, but that view doesn't provide attribute keys? I would like for the advisory parameter to be used in this case, otherwise we surprisingly explode the user's metric cardinality in Java Instrumentation (where we rely heavily on the attribute keys advisory parameter) 😢 https://github.com/open-telemetry/opentelemetry-java/issues/7134#issuecomment-2679529209

Reviewer 3: thanks! also @jack-berg mentioned it's already covered [here](https://github.com/open-telemetry/opentelemetry-specification/blame/6f6559fd5bfe26f97599e56c22c6d9a83e268722/specification/metrics/sdk.md#L366-L372):

>  Users can provide `attribute_keys`, but it is up to their discretion.
  Therefore, the stream configuration parameter needs to be structured to
  accept `attribute_keys`, but MUST NOT obligate a user to provide them.
  If the user does not provide any value, the SDK SHOULD use
  the [`Attributes`](./api.md#instrument-advisory-parameters) advisory
  parameter configured on the instrument instead. If the `Attributes`
  advisory parameter is absent, all attributes MUST be kept.

first-seen advisory parameters and log an appropriate error as described in
[duplicate instrument registrations](#duplicate-instrument-registration).

> If a [View](#view) is configured with the same settings as advisory parameters,

Reviewer 2: What does "the same settings" mean?

Author: good point, let me clarify better.

Author: Modified. Please check https://github.com/open-telemetry/opentelemetry-specification/pull/4389/commits/daabd2d42163f2c308c2b4f4a002660a8a559f10

the View MUST take precedence over the advisory parameters.

> #### Instrument advisory parameter: `ExplicitBucketBoundaries`

Reviewer 7: We generally haven't explicitly documented the API methods or parameters in the SDK specification.  It is already stated that the SDK is an implementation of the API.  I think the general statement above "If a View is configured with the same settings as advisory parameters,
the View MUST take precedence over the advisory parameters." is already enough to clarify the precedence ordering.

I don't think this section, or the one below, is necessary unless is is clarifying something additional (other than precedence or what is already in the API).

Author: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-advisory-parameters The API doc says "OpenTelemetry SDKs MUST handle advisory parameters as described [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#instrument-advisory-parameters)", and that here is what this section.

I am okay with just keeping the general statement about view takes precedency alone and remove rest. Will see if there are other feedback on this. I have seen recently that we try to be more explicit in SDK specs (eg: Logger.Enabled), but does not necessarily have to do it for advisory.

Author: Resolving. Please re-open if there is changes requested.
